nntrainer/dataset/batch_queue.h

```diff
@@ -91,11 +96,184 @@ class BatchQueue {
   std::condition_variable_any q_reader_cv;
   std::condition_variable_any q_writer_cv;
 
+  std::queue<std::unique_ptr<T>> q;
+};
+
+/**
+ * @brief A view container that calls a callback on destruct
+ * @note the callback must be noexcept, and the given underlying data must
+ * outleave the lifetime of this class
+ *
+ * @tparam T underlying type
+ */
+template <typename T> class ScopedView {
+public:
   /**
-   * @todo consider using circular buffer if this is too slow
+   * @brief Construct a new Scoped View object
    *
+   * @param data_ reference of the underlying data
+   * @param on_notify_ callback to be called on exit, this is not copied but
+   * reused
    */
-  std::queue<std::unique_ptr<T>> q;
+  ScopedView(T *data_, std::function<void(void)> &&on_notify_) :
+    data(data_),
+    on_notify(std::forward<std::function<void(void)>>(on_notify_)) {}
+
+  ScopedView(const ScopedView &rhs) = delete;
+  ScopedView &operator=(const ScopedView &rhs) = delete;
+
+  ScopedView(ScopedView &&rhs) = default;
+  ScopedView &operator=(ScopedView &&rhs) = default;
+
+  /**
+   * @brief Destroy the Scoped View object, callback is called at this time
+   *
+   */
+  ~ScopedView() {
+    try {
+      on_notify();
+    } catch (...) {
+      ml_loge("while notifiying, error happened");
+    }
+  }
+
+  /**
+   * @brief get the underlying data
+   *
+   * @return T & reference to the underlying data
+   */
+  T &get() { return *data; }
+
+private:
+  T *data;
```

Comment:
please add some description for class members.

Reply:
Added